File: apps/shared/Sources/FerniShared/Models/VoiceState.swift
import Foundation

// MARK: - Voice State

/// Represents the current state of the voice session
public enum VoiceState: Equatable {
    case disconnected
    case connecting
    case connected
    case listening
    case speaking
    case thinking
    case error(String)

    /// User-friendly title for the state
    public var title: String {
        switch self {
        case .disconnected: return "Ready"
        case .connecting: return "Connecting..."
        case .connected: return "Connected"
        case .listening: return "Listening"
        case .speaking: return "Speaking"
        case .thinking: return "Thinking"
        case .error: return "Connection Issue"
        }
    }

    /// Whether the session is active (not disconnected or error)
    public var isActive: Bool {
        switch self {
        case .disconnected, .error:
            return false
        default:
            return true
        }
    }

    /// Whether audio visualization should be shown
    public var showWaveform: Bool {
        switch self {
        case .connected, .listening, .speaking:
            return true
        default:
            return false
        }
    }

    /// Breathing animation intensity multiplier
    /// NOTE: All active states use same intensity - no listening/speaking distinction
    public var breathingIntensity: Double {
        switch self {
        case .connected, .listening, .speaking:
            // All active conversation states use same intensity
            return 1.0
        case .thinking:
            return 0.8
        default:
            return 0.6
        }
    }
}

// MARK: - Voice Event

/// Events emitted during voice session
public enum VoiceEvent {
    case stateChanged(VoiceState)
    case transcription(text: String, isAgent: Bool, isFinal: Bool)
    case handoff(from: String, to: String)
    case audioLevel(Float)
    case error(Error)
}

// MARK: - Audio Level Sample

/// Audio level sample for waveform visualization
public struct AudioLevelSample: Identifiable {
    public let id = UUID()
    public let level: Float
    public let timestamp: Date

    public init(level: Float) {
        self.level = max(0, min(1, level))
        self.timestamp = Date()
    }
}

File: apps/android-native/app/src/main/java/com/ferni/voice/models/VoiceState.kt
package com.ferni.voice.models

/**
 * Represents the current state of the voice session.
 */
sealed class VoiceState {
    data object Disconnected : VoiceState()
    data object Connecting : VoiceState()
    data object Connected : VoiceState()
    data object Listening : VoiceState()
    data object Speaking : VoiceState()
    data object Thinking : VoiceState()
    data class Error(val message: String) : VoiceState()

    /** User-friendly title for the state */
    val title: String
        get() = when (this) {
            is Disconnected -> "Ready"
            is Connecting -> "Connecting..."
            is Connected -> "Connected"
            is Listening -> "Listening"
            is Speaking -> "Speaking"
            is Thinking -> "Thinking"
            is Error -> "Connection Issue"
        }

    /** Whether the session is active (not disconnected or error) */
    val isActive: Boolean
        get() = when (this) {
            is Disconnected, is Error -> false
            else -> true
        }

    /** Whether audio visualization should be shown */
    val showWaveform: Boolean
        get() = when (this) {
            is Connected, is Listening, is Speaking -> true
            else -> false
        }

    /**
     * Breathing animation intensity multiplier.
     * All active states use same intensity - no listening/speaking distinction.
     */
    val breathingIntensity: Float
        get() = when (this) {
            is Connected, is Listening, is Speaking -> 1.0f
            is Thinking -> 0.8f
            else -> 0.6f
        }
}

/**
 * Events emitted during voice session.
 */
sealed class VoiceEvent {
    data class StateChanged(val state: VoiceState) : VoiceEvent()
    data class Transcription(
        val text: String,
        val isAgent: Boolean,
        val isFinal: Boolean
    ) : VoiceEvent()
    data class Handoff(val from: String, val to: String) : VoiceEvent()
    data class AudioLevel(val level: Float) : VoiceEvent()
    data class ErrorEvent(val error: Throwable) : VoiceEvent()
}

/**
 * Emotion hints for avatar expressions.
 *
 * These values come from the backend and are used by:
 * - LiveKitSession: parses emotion_event data messages
 * - VoiceViewModel: maps hints to personality animations + micro-expressions
 * - BetterThanHumanEngine: displays emotional state in avatar
 */
enum class EmotionHint {
    HAPPY,
    EXCITED,
    CURIOUS,
    EMPATHETIC,
    ENCOURAGING,
    NEUTRAL,
    THINKING,  // Contemplative, processing state
    CALM       // Peaceful, relaxed state
}
